File: backend/app/ai/schemas/ai.py
"""AI Pydantic schemas — request/response models for the AI layer."""

from datetime import datetime

from pydantic import BaseModel, Field

# ── Context request / response ─────────────────────────────────────────────────


class ContextRequest(BaseModel):
    universe_id: str = Field(
        ..., min_length=1, description="UUID of the universe to contextualise."
    )


class UniverseSnippet(BaseModel):
    id: str
    name: str
    genre: str
    description: str | None
    tone: str | None
    status: str


class CharacterSnippet(BaseModel):
    id: str
    name: str
    role: str | None
    biography: str | None
    personality: str | None
    goals: str | None
    motivations: str | None
    strengths: str | None
    weaknesses: str | None


class LocationSnippet(BaseModel):
    id: str
    name: str
    type: str | None
    description: str | None
    climate: str | None
    culture: str | None


class OrganizationSnippet(BaseModel):
    id: str
    name: str
    type: str | None
    description: str | None
    leader: str | None
    purpose: str | None


class WorldObjectSnippet(BaseModel):
    id: str
    name: str
    category: str | None
    description: str | None
    origin: str | None
    abilities: str | None


class WorldRuleSnippet(BaseModel):
    id: str
    title: str
    category: str | None
    description: str | None
    limitations: str | None
    exceptions: str | None


class RelationshipSnippet(BaseModel):
    id: str
    source: str  # resolved display name or entity_id fallback
    source_type: str
    relationship: str  # relationship_type value
    target: str  # resolved display name or entity_id fallback
    target_type: str
    strength: int | None
    direction: str
    description: str | None


class TimelineEventSnippet(BaseModel):
    id: str
    title: str
    event_type: str
    status: str
    start_date: str | None
    end_date: str | None
    importance: int | None
    description: str | None
    # Participant summary: "EntityType:EntityId (role)" strings
    participants: list[str] = []


class ContextMetadata(BaseModel):
    generated_at: datetime
    counts: dict[str, int]
    version: str = "1.0"


class UniverseContext(BaseModel):
    """Complete AI-ready representation of a universe."""

    universe: UniverseSnippet
    characters: list[CharacterSnippet]
    locations: list[LocationSnippet]
    organizations: list[OrganizationSnippet]
    objects: list[WorldObjectSnippet]
    world_rules: list[WorldRuleSnippet]
    relationships: list[RelationshipSnippet] = []
    timeline: list[TimelineEventSnippet] = []
    metadata: ContextMetadata


# ── Provider health ────────────────────────────────────────────────────────────


class ProviderHealthResponse(BaseModel):
    provider_name: str
    model: str
    healthy: bool
    message: str
    version: str | None = None


# ── Generation (future use) ────────────────────────────────────────────────────


class GenerationRequest(BaseModel):
    universe_id: str
    prompt_key: str = Field(
        ...,
        description="Prompt template key, e.g. 'universe_summary' or 'character_analysis'.",
    )
    user_question: str | None = None
    temperature: float | None = Field(None, ge=0.0, le=2.0)
    max_tokens: int | None = Field(None, ge=1, le=8192)


class GenerationResponse(BaseModel):
    text: str
    model: str
    provider: str
    prompt_tokens: int | None = None
    completion_tokens: int | None = None

File: frontend/types/ai.ts
/**
 * AI TypeScript types — mirrors backend AI schemas.
 */

// ── Universe Context ──────────────────────────────────────────────────────────

export interface UniverseSnippet {
  id: string;
  name: string;
  genre: string;
  description: string | null;
  tone: string | null;
  status: string;
}

export interface CharacterSnippet {
  id: string;
  name: string;
  role: string | null;
  biography: string | null;
  personality: string | null;
  goals: string | null;
  motivations: string | null;
  strengths: string | null;
  weaknesses: string | null;
}

export interface LocationSnippet {
  id: string;
  name: string;
  type: string | null;
  description: string | null;
  climate: string | null;
  culture: string | null;
}

export interface OrganizationSnippet {
  id: string;
  name: string;
  type: string | null;
  description: string | null;
  leader: string | null;
  purpose: string | null;
}

export interface WorldObjectSnippet {
  id: string;
  name: string;
  category: string | null;
  description: string | null;
  origin: string | null;
  abilities: string | null;
}

export interface WorldRuleSnippet {
  id: string;
  title: string;
  category: string | null;
  description: string | null;
  limitations: string | null;
  exceptions: string | null;
}

export interface RelationshipSnippet {
  id: string;
  source: string;
  source_type: string;
  relationship: string;
  target: string;
  target_type: string;
  strength: number | null;
  direction: string;
  description: string | null;
}

export interface ContextMetadata {
  generated_at: string;
  counts: {
    characters: number;
    locations: number;
    organizations: number;
    objects: number;
    world_rules: number;
  };
  version: string;
}

export interface UniverseContext {
  universe: UniverseSnippet;
  characters: CharacterSnippet[];
  locations: LocationSnippet[];
  organizations: OrganizationSnippet[];
  objects: WorldObjectSnippet[];
  world_rules: WorldRuleSnippet[];
  relationships: RelationshipSnippet[];
  metadata: ContextMetadata;
}

// ── Provider health ────────────────────────────────────────────────────────────

export interface ProviderHealth {
  provider_name: string;
  model: string;
  healthy: boolean;
  message: string;
  version: string | null;
}

// ── Request payloads ───────────────────────────────────────────────────────────

export interface ContextRequest {
  universe_id: string;
}
